sources/api/migration/migrations/v1.12.0/k8s-private-pki-path/src/main.rs

```diff
@@ -0,0 +1,81 @@
+#![deny(rust_2018_idioms)]
+
+use migration_helpers::{migrate, Migration, MigrationData, Result};
+use std::process;
+
+const SETTING: &'static str = "configuration-files.kubelet-server-key.path";
+const OLD_VALUE: &'static str = "/etc/kubernetes/pki/kubelet-server.key";
+const NEW_VALUE: &'static str = "/etc/kubernetes/pki/private/kubelet-server.key";
+
+/// We moved the render output location for the kubelet PKI private key to be in a restricted
+/// subdirectory. We need to update this output path in the stored configuration so updated nodes
+/// pick up the change.
+fn run() -> Result<()> {
+    migrate(KubeletServerKey {})
+}
+
+pub struct KubeletServerKey {}
+
+impl KubeletServerKey {
+    fn migrate(&mut self, mut input: MigrationData, action: &'static str) -> Result<MigrationData> {
+        let old_value;
+        let new_value;
+        if action == "upgrade" {
+            old_value = OLD_VALUE;
+            new_value = NEW_VALUE;
+        } else {
+            // Downgrade: everything old is new again
+            old_value = NEW_VALUE;
+            new_value = OLD_VALUE;
```

Comment:
If we know the old value is bad and shouldn't be used, do we still want to replace the good value with the old value on downgrade? Maybe there's a case for not doing anything on downgrade.

Reply:
If we're downgrading then the older version (v1.11.1) is going to expect the file to be in the old location. Or do I have a fundamental misunderstanding of what a downgrade is?

Reply:
Ah you're right. I missed that part. Kubelet will be looking for the file at that exact path. Ignore me!